node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -0,0 +1,240 @@
+}
+
+/**
+  * This API is totally based on how RevVault.rho is written. If the `RevVault.rho` is re-written or changed,
+  * this API might end up with useless.
+  */
+final case class TransactionAPIImpl[F[_]: Concurrent](
+    blockReportAPI: BlockReportAPI[F],
+    // The transferUnforgeable can be retrieved based on the deployer and the timestamp of RevVault.rho
+    // in the genesis ceremony.
+    transferUnforgeable: Par
+) extends TransactionAPI[F] {
+  // Helper which accepts TransactionType constructor
+  private def makeDeployTransaction(
+      d: DeployInfoWithEventData,
+      report: SingleReport,
+      txCtor: String => TransactionType
+  ) = findTransactions(report).map(t => TransactionInfo(t, txCtor(d.deployInfo.get.sig)))
+
+  def getTransaction(blockHash: BlockHash): F[Seq[TransactionInfo]] = {
+    val blockHashStr = Base16.encode(blockHash.toByteArray)
+    blockReportAPI.blockReport(blockHash, forceReplay = false).map {
+      case Right(b) =>
+        val userTransactions = b.deploys.flatMap(d => {
+          d.report
+            .zip(Seq(PreCharge, UserDeploy, Refund))
+            .flatMap {
+              case (report, txCtor) =>
+                makeDeployTransaction(d, report, txCtor)
+            }
+        })
+        val systemDeployTransactions = b.systemDeploys.flatMap { s =>
+          // system doesn't get precharge and refund , so it would always get one
+          val transactions = findTransactions(s.report.head)
+          s.systemDeploy.get.systemDeploy.value match {
+            case SlashSystemDeployDataProto(_, _) =>
+              transactions.map(
+                t => TransactionInfo(t, SlashingDeploy(blockHashStr))
+              )
+            case CloseBlockSystemDeployDataProto() =>
+              transactions.map(t => TransactionInfo(t, CloseBlock(blockHashStr)))
+          }
+        }
+        userTransactions ++ systemDeployTransactions
+
+      case Left(_) => Seq.empty
+    }
+  }
+
+  private def findTransactions(report: SingleReport): Seq[Transaction] = {
+    val transactions = report.events
+      .map(_.report.value)
+      .collect {
+        case ReportCommProto(Some(consume), produces) =>
+          consume.channels.headOption.map((_, produces))
+      }
+      .flatten
+      .collect { case (channel, produces) if channel == transferUnforgeable => produces.head }
+      .map { produce =>
+        val fromAddr       = produce.data.get.pars.head.exprs.head.getGString
+        val toAddr         = produce.data.get.pars(2).exprs.head.getGString
+        val amount         = produce.data.get.pars(3).exprs.head.getGInt
+        val retUnforgeable = produce.data.get.pars(5)
+        Transaction(fromAddr, toAddr, amount, retUnforgeable, None)
+      }
+
+    // traverse all the transactions to check if the transaction is completed or failed with reason
+    transactions.map { transaction =>
+      report.events
+        .find { p =>
```

Comment:
This is another place where events will be iterated for each transaction.

In these kind of situations, first build the structure for quick access by the key you want to find (Map) or build a collection which you can join (like SQL inner join).

```scala
val failedMap = report.events.iterator
  .map(_.report.value)
  .collect {
    case ReportProduceProto(Some(channel), data) =>
      val success =
        data.get.pars.head.exprs.head.getETupleBody.ps.head.exprs.head.getGBool
      val failedReason =
        if (success) None
        else Some(data.get.pars.head.exprs.head.getETupleBody.ps(1).exprs.head.getGString)
      (channel, failedReason)
  }
  .toMap
```
This map can then be used directly when Transaction is constructed.

Reply:
I wrote a lot to reply to you. But find that you are right. Haha. Good work.